Added arbitrary second order phase space expectation/variances

josh146 opened the pull request:
**Description of the Change:**

* Added the method `BaseState.poly_quad_expectations(A, d, k, phi=0)`. This returns the expectation `E[P(r)]` and variance `var[P(r)]` of arbitrary second order polynomials of the form `P(r) = r^T A r/2 + r^T d+kI`, where the vector of quadrature operators  `r = (x1, x2, ..., x_n, p1, ..., p_n)` has been optionally rotated by angle `phi` in the phase space. In particular,
  - `A` is a  2Nx2N real symmetric matrix containing the quadratic coefficients,
  - `d` is a length-2N real vector containing the linear coefficients, and
  - `k` is the constant term.

  In the Gaussian backend, the formulas for the [quadratic form](https://en.wikipedia.org/wiki/Quadratic_form_(statistics)) are used. In the Fock backends, `E[P(r)] = Tr(P(r)*rho)` is calculated directly using `np.einsum`.

* Added unit tests for `poly_quad_expectations`. These test that, for `E[x]` and `E[p]`, that the result agrees with `quad_expectations`. In addition, it also tests against the known result for `E[x^2]`, `E[p^2]`, `E[xp]`, as well as various arbitrary quadratic forms - both single mode, and multi mode.

* `BaseState.mean_photon` now returns a tuple of both the mean and the variance, bringing it in line with `BaseState.quad_expectations`.

* Modified the `GaussianState.mean_photon` method to use the equations

  `<n> = [Tr(cov)+muT.mu]/(2\hbar) - 1/2`
  `\sigma_n^2 = [Tr(cov^2)+2muT.cov.mu]/(2\hbar^2) - 1/4`

* Added additional unit tests for `mean_photon`, expanding to displaced squeezed state, and also testing the variance.

**Benefits:**

* Added the ability to return arbitrary second order expectations

* The output of `mean_photon` now matches those of the other expectation methods

* Calculation of the mean photon number is now much more efficient and accurate for Gaussian states

**Possible Drawbacks:**

* At the moment, `poly_quad_expectations` does not work on the TF backend in batch mode - this will need to be subclassed and fixed within the `TFFockState` class. In addition, symbolic evaluation will also fail until the NumPy function calls are replaced with their TF equivalent.

* In the Fock backends:
  - the `poly_quad_expectations` calculation scales quite badly as the number of operators acting on _different_ modes increases. This is due to the increasing number of dimensions required in the einsum calculation. This can be somewhat alleviated by reducing the Fock cutoff, at the expense of accuracy. Alternatively, the use of sparse matrices might solve this problem.
  - There is a current limit of ~15 modes in the multimode expectation, due to the use of `einsum`.

**Related GitHub Issues:**

n/a

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -359,6 +362,56 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+
+        * :math:`A\in\mathbb{R}^{2N\times 2N}` is a symmetric matrix
+          representing the quadratic coefficients,
+        * :math:`\mathbf{d}\in\mathbb{R}^{2N}` is a real vector representing
```
In the PR overview, it says that d is a 2N-dimensional complex vector, not real. Just to confirm, it should be real?

josh146 replied:
Yes I guess it should be, otherwise the observable would be non-Hermitian

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -359,6 +362,56 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+          the linear coefficients,
+        * :math:`k\in\mathbb{R}` represents the constant term, and
+        * :math:`\mathbf{r} = (\x_1,\dots,\x_N,\p_1,\dots,\p_N)` is the vector
+          of quadrature operators in :math:`xp`-ordering.
```
Just curious, what would it be called if we used the ordering (x1,p1,x2,p2,...)?

josh146 replied:
Good question, @nquesada? I just copied and pasted this from somewhere else in the SF codebase. I've seen one called the symmetric ordering, and one called the `xp`-ordering.

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -359,6 +362,56 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+
+        as well as the variance
+
+        .. math:: \Delta P(\mathbf{r})^2 = \langle P(\mathbf{r})^2\rangle - \braket{P(\mathbf{r})}^2
```
not sure how this renders, but how come we are mixing `\langle\rangle` and `\braket`?

josh146 replied:
Ah, I'll fix that. `\braket` is just a macro for `\langle#1\rangle`

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -359,6 +362,56 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+        Args:
+            A (array): a real symmetric 2Nx2N NumPy array, representing the quadratic
+                coefficients of the second order quadrature polynomial.
+            d (array): a real symmetric length-2N NumPy array, representing the linear
```
symmetric?

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -359,6 +362,56 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+            A (array): a real symmetric 2Nx2N NumPy array, representing the quadratic
+                coefficients of the second order quadrature polynomial.
+            d (array): a real symmetric length-2N NumPy array, representing the linear
+                coefficients of the second order quadrature polynomial. Default the zero vector.
```
typo: 'Default is the zero vector' or 'Defaults to the zero vector'

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -359,6 +362,56 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+                coefficients of the second order quadrature polynomial. Default the zero vector.
+            k (float): the constant term. Default 0.
+            phi (float): quadrature angle, clockwise from the positive :math:`x` axis. If provided,
+                the vectori of quadrature operators :math:`\mathbf{r}` is first rotated
```
typo: 'vectori'

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -359,6 +362,56 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+                by angle :math:`\phi` in the phase space.
+
+        Keyword args:
+            worksize (int): For Fock backends, the additional Fock basis truncation used to
```
I think we can fix the worksize, no? For instance, truncation of the x^2 operator causes only the last matrix element to be inaccurate. I think @nquesada can probably tell you hard bounds on the maximal worksize needed

josh146 replied:
@nquesada:
> worksize of cutoff+1 should be fine

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -689,6 +744,124 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+            raise ValueError("Matrix of quadratic coefficients A must be of size 2Nx2N.")
+
+        if np.any(A.T != A):
+            raise ValueError("Matrix of quadratic coefficients A must be symmetric.")
```
We should probably use `isclose` here instead, because these matrices may come downstream from some other numerical computation and may not be *exactly* symmetric

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -689,6 +744,124 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+            linear_coeff = np.zeros([2*self._modes])
+        else:
+            linear_coeff = d.copy()
+            linear_coeff[self._modes:] = -d[self._modes:]
```
Couldn't understand what this line is for

josh146 replied:
This actually really confuses me - for some reason, the einsum Fock calculation doesn't provide the same answers as the Guassian backend/theoretical result _unless_ I first modify `d` and `A` such that

1. The linear coefficients on the momentum operators are negated: `d' = (d_1,...,d_n,-d_{n+1},...,-d_{2n})`, and

2. The block-off diagonal elements of `A` (i.e. so the coefficients of the `x_np_m` terms) are also negated.

The only reason I know that this works is because I had the _exact_ same issue in OpenBoson, and this is the trick that allowed the Fock backends to provide the correct result.

co9olguy replied:
Perhaps @nquesada knows why there is a discrepency between backends? Is this something that the SF frontend compensates for?

josh146 replied:
@nquesada didn't know either

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -689,6 +744,124 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+        dim = self._cutoff + worksize
+
+        # construct the x and p operators
+        a = np.diag(np.sqrt(np.arange(1, dim)), 1)
```
can we implement these as sparse matrices? Might give significant savings

co9olguy replied:
Let's remember to do it before bumping the release number

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -689,6 +744,124 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+        # reduce the size of A so that we only consider modes
+        # which we need to calculate the expectation value for
+        rows = ex_modes + [i+self._modes for i in ex_modes]
+        quad_coeffs = A[:, rows][rows]/2
```
what is the advantage of putting the 1/2 in the convention for the quadratic form? This forces users to compensate for it in A, which they might forget to do...

josh146 replied:
What are your thoughts on changing it?

co9olguy replied:
I'd prefer to not have the 1/2 in the convention. This would likely cause less surprises

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -689,6 +744,124 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+        rows = ex_modes + [i+self._modes for i in ex_modes]
+        quad_coeffs = A[:, rows][rows]/2
+        quad_coeffs[num_modes:, :num_modes] = -quad_coeffs[num_modes:, :num_modes]
+        quad_coeffs[:num_modes, num_modes:] = -quad_coeffs[:num_modes, num_modes:]
```
See question above. What is the role of these lines? Why do we make these particular elements of the matrix negative?

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -689,6 +744,124 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+        ind3 = ''.join([str(i)+str(j) for i, j in zip(ind1[1::2], ind2[2::2])])
+        ind = "{},{}->{}".format(ind1, ind2, ind3)
+
+        if np.all(quad_coeffs == 0.):
```
Should we use `isclose` here as well?

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -689,6 +744,124 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+
+        # compute the polynomial
+        # For 3 modes, this gives the einsum (with brackets denoting modes):
+        # 'a(bc)(de)(fg),a(ch)(ei)(gj)->(bh)(di)(fj)' applied to r, A@r
```
Just for my own understanding, what does index `a` correspond to?

josh146 replied:
For 3 modes (for example), array `r` has dimensions `aijklmn`. `a` corresponds to the index in the vector of quadrature operators `r = (x_1,...,x_n,p_1,...,p_n)`, and the remaining indices `ijklmn` are the elements of the operator acting on a 3 mode density matrix.

josh146 replied:
So, in effect, matrix of quadratic coefficients `A` acts only on index `a`, this index is then summed, and then each mode of `r`, `A@r` undergoes matrix multiplication

co9olguy replied:
Makes sense. Maybe add this short explanation to the code comment directly?

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -917,10 +1090,55 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+        if A.shape != (2*self._modes, 2*self._modes):
+            raise ValueError("Matrix of quadratic coefficients A must be of size 2Nx2N.")
+
+        if np.any(A.T != A):
```
`isclose`

co9olguy commented on strawberryfields/backends/states.py:
```diff
@@ -917,10 +1090,55 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+        # expectation value E[P(r)] = tr(A.cov) + muT.A.mu + muT.d + k
+        mean = np.trace(H @ cov) + k2
+        # variance Var[P(r)] = 2tr(A.cov.A.cov) + 4*muT.A.cov.A.mu + dT.cov.d
+        var = 2*np.trace(H @ cov @ H @ cov) + d2.T @ cov @ d2
```
equations in comments don't obviously appear to match implementation

co9olguy commented on strawberryfields/backends/tfbackend/states.py:
```diff
@@ -501,14 +501,20 @@ def mean_photon(self, mode, **kwargs):
             flat_rho = tf.reshape(rho, [-1, self.cutoff_dim ** 2])
 
             nbar = tf.real(tf.reduce_sum(flat_rho * flat_n, axis=1)) # implements a batched tr(rho * n)
+            nbarSq = tf.real(tf.reduce_sum(flat_rho * flat_n**2, axis=1)) # implements a batched tr(rho * n)
```
comment should have `n**2`

co9olguy commented on strawberryfields/backends/tfbackend/states.py:
```diff
@@ -608,6 +614,64 @@ def wigner(self, mode, xvec, pvec):
+            A (array): a real symmetric 2Nx2N NumPy array, representing the quadratic
+                coefficients of the second order quadrature polynomial.
+            d (array): a real symmetric length-2N NumPy array, representing the linear
+                coefficients of the second order quadrature polynomial. Default the zero vector.
```
adjust 'Default the zero vector' to whatever we change this sentence to elsewhere

co9olguy commented on strawberryfields/backends/tfbackend/states.py:
```diff
@@ -608,6 +614,64 @@ def wigner(self, mode, xvec, pvec):
+            tuple (float, float): expectation value and variance
+        """
+        if self._eval and not self.batched:
+            return super().poly_quad_expectation(A, d, k, phi, **kwargs)
```
This might cause confusion in the cas that `self._eval==True` and `self.batched==True`, but where the user passes `eval=False` or a session within `kwargs` when they call this method. In that case, the class should override the default `self._eval`.

For now, we can catch this by calling checking `ops.check_for_eval(kwargs)`

josh146 replied:
:+1: Do you want to make this change? It will probably also need to be changed in the `TFFockState.wigner` method.

co9olguy replied:
Will do. 
The `wigner` method is ok, since it explicitly states both in the documentation and an exception that it only supports that case

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
         self.assertAllAlmostEqual(res.flatten(), res_exact.flatten(), delta=self.tol)
 
 
+class PolyQuadExpectationSingleMode(BaseTest):
```
Just noticed this now, but the test classes in `test_states.py` all have names without `Test` at the end, while in all other test files (that I'm aware of), we use the convention `SomeKindOfTest(ParentTest)`

josh146 replied:
Let's leave this for a future PR, and perhaps make an issue for it?

co9olguy replied:
This was quick, just took care of it

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
 
+class PolyQuadExpectationSingleMode(BaseTest):
+    num_subsystems = 3
+    cutoff = 12
```
We should prefer not to have one of the user-adjustable test parameters explicitly bypassed in the tests (at least, not without warning).

josh146 replied:
I originally tried that; unfortunately, a cutoff of 5 (the default) causes the expectation of the quadratic terms to have too much error, causing the assertions to fail

co9olguy replied:
We should either change the parameters of the test to fit within that cutoff, or add a warning msg inside the test that the user-provided value is overridden (if it is lower than 12)

co9olguy replied:
I added a logger warning to the relevant tests

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        Args:
+            mu (array): means vector
+            cov (array): covariance matrix
+            func (function): function acting on the random variables X, P, XP,
```
only X, P, and XP?

josh146 replied:
Because you can get X^2 from X, etc. :laughing: Maybe I was a bit lazy there

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+
+        mean, var = state.poly_quad_expectation(A, d, k)
+        self.assertEqual(mean, 0)
+        self.assertEqual(var, 0)
```
We probably want to set these to `assertAlmostEqual` if we change the backend implementations to use `np.isclose`

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        mean, var = state.poly_quad_expectation(A, d, k, phi=self.qphi)
+
+        self.assertAlmostEqual(mean, self.mu[0], delta=self.tol)
+        self.assertAlmostEqual(var, self.cov[0, 0], delta=self.tol)
```
Are these expected results independent of the `hbar` value?

josh146 replied:
Yep, self.mu and self.cov are calculated taking hbar into account

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        mean, var = state.poly_quad_expectation(A, d, k, phi=self.qphi)
+
+        self.assertAlmostEqual(mean, self.mu[1], delta=self.tol)
+        self.assertAlmostEqual(var, self.cov[1, 1], delta=self.tol)
```
same question about `hbar`

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        state = self.circuit.state()
+        mean, var = state.poly_quad_expectation(A, d, k, phi=self.qphi)
+
+        # E(2x+p) = 2E(x) + E(p)
```
update this comment to use `a` and `b`

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        mean_expected = d[0]*self.mu[0] + d[3]*self.mu[1]
+        self.assertAlmostEqual(mean, mean_expected, delta=self.tol)
+
+        # var(2x+p) = 4var(x)+var(p)+4cov(x,p)
```
same

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+
+        # set quadratic coefficient
+        A = np.zeros([6, 6])
+        A[3, 0] = A[0, 3] = 1
```
This is more `xp+px` than `xp`, no?

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+
+        # construct the vector of means and covariance matrix for each mode
+        # note that we use (x0,p0,x1,p1,...) ordering here.
+        for i, (x, y, z) in enumerate(zip(self.a_list, self.r_list, self.phi_list)):
```
Might help for code clarify to use `i, (a, r, phi)`

josh146 replied:
I originally had that, unfortunately the linter complains that we are modifying global variables

co9olguy replied:
`(a_, r_, phi_)`?

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        S1 = block_diag(BS, np.identity(2))
+
+        if self.linear_modes == 3:
+            S2 = block_diag(np.identity(2), BS)
```
unused? (same question applies throughout)

josh146 replied:
See https://github.com/XanaduAI/strawberryfields/pull/34#discussion_r221158250

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        self.circuit.reset(pure=self.kwargs['pure'])
+
+        # squeeze and displace each mode
+        for i, (x, y, z) in enumerate(zip(self.a_list, self.r_list, self.phi_list)):
```
see comment above

josh146 replied:
see https://github.com/XanaduAI/strawberryfields/pull/34#discussion_r221158339

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        self.cov = S2 @ S1 @ cov @ S1.T @ S2.T
+
+    def test_multi_mode_linear_combination(self):
+        """Test that the correct result is returned for E(2x0 + x1 - p0)"""
```
docstring should be changed to reflect actual parameters used

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        state = self.circuit.state()
+        mean, var = state.poly_quad_expectation(A, d, k, phi=self.qphi)
+
+        # E(2x0 + x1 - p0) = 2E(x0) + E(x1) - E(p0)
```
update comment to reflect parameter values

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        mean_expected = np.array([x0, p0, x1, 0, x2, 0]) @ self.mu
+        self.assertAlmostEqual(mean, mean_expected, delta=self.tol)
+
+        # var(x0 + x1 + p0) = 4var(x0)+var(x1)+var(p0)-4cov(x0,p0) + 4cov(x0, x1) - 2*cov(x1, p0)
```
update

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        self.assertAlmostEqual(var, var_ex, delta=self.tol)
+
+    def test_xp(self):
+        """Test that the correct result is returned for E(xp)"""
```
Since we force the polynomials to be symmetric, we shouldn't be testing non-symmetric operators

josh146 replied:
Bit of lazy writing on my half. The expectation value of the polynomial xp in the phase space is identical/equivalent to the expectation value of the operator (xp+px) after Weyl quantization

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        self.assertAlmostEqual(var, var_expected, delta=self.tol)
+
+    def test_multi_mode_quadratic_combination(self):
+        """Test that the correct result is returned for E(x0x1+x0p1)"""
```
See earlier comment about testing only symmetric operators

josh146 replied:
Same as above; in this case, E(x0x1+x0p1) is the polynomial in the phase space, not the quantized operators

co9olguy commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        self.circuit.reset(pure=self.kwargs['pure'])
+
+        # squeeze and displace each mode
+        for i, (x, y, z) in enumerate(zip(self.a_list, self.r_list, self.phi_list)):
```
see above comment about variable names

josh146 replied:
See https://github.com/XanaduAI/strawberryfields/pull/34#discussion_r221158339

josh146 commented on strawberryfields/backends/states.py:
```diff
@@ -689,6 +744,124 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+        if phi != 0:
+            # rotate the quadrature operators
+            xold = x.copy()
+            pold = p.copy()
```
Only because we have simultaneous linear equations in the next two lines, using the same variable name. Since `x` and `p` are NumPy arrays, using temporary variables does not work, as it provides a view to the same array.

co9olguy replied:
What if we just use different names for the transformed operators? Then we wouldn't need to copy, right?

```
x_ = np.sqrt(...)
p_ = -1j * ...

if phi != 0:
    x = np.cos(phi) * x - np.sin(phi) * p
    p = np.sin(phi) * x + np.sin(phi) * p
else:
    x = x_
    p = p_

josh146 commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+        A[3, 3] = 2*c1
+
+        A[3, 0] = c2
+        A[0, 3] = c2
```
The factor of 1/2 only need to be accounted for on the diagonal of A, as the off-diagonals are counted twice by default

josh146 commented on tests/test_states.py:
```diff
@@ -220,6 +247,394 @@ def test_squeezed_coherent(self):
+    # define circuit parameters
+    a_list = [0.344+0.123j, 0.432+0.123j, 0.1231-0.543j][:linear_modes]
+    r_list = [0.1065, 0.32, 0.4123][:linear_modes]
+    phi_list = [0.897, 0.31, 0.9643][:linear_modes]
```
That's correct; the `linear_modes=3` case I added just as a sanity check, to ensure that this works locally. The only reason I didn't include it as a separate test is that it takes quite long to calculate on the Fock backend, approximately 30s.

co9olguy replied:
k then let's remove from tests

josh146 commented on strawberryfields/backends/tfbackend/states.py:
```diff
@@ -550,7 +556,7 @@ def dm(self, **kwargs):
                   which is specified at initialization.
                 * A Tensorflow Session or feed_dict may also be passed via the keys ``session`` or ``feed_dict``, respectively.
-                  These are ignored when ``eval=False``; when ``eval=True``, they are used when numerically evaluating the underlying quantity.
+                  If a Session is supplied, then `eval` is overriden and the numerical evaluation takes place in the provided Session.
```
Should be ``eval``

josh146 commented:
Update: added additional `mean_photon` tests (as there were bugs in the previous implementation that weren't picked up from the tests), and added an additional `poly_quad_expectation` test to ensure that `E(n)` and `var(n)` match that of `mean_photon`.

co9olguy commented:
Not sure why the codecov stuff failed, but tests are passing, so I'm now merging
